testing/opae-c/test_feature_enum_c.cpp

```diff
@@ -0,0 +1,257 @@
+#include <opae/fpga.h>
+#include <uuid/uuid.h>
+#include "opae_int.h"
+#include "safe_string/safe_string.h"
+
+#include "types_int.h"
+#include "xfpga.h"
+#include "intel-fpga.h"
+#include "feature_pluginmgr.h"
+
+#ifdef __cplusplus
+}
+#endif
+
+#include <array>
+#include <cstdlib>
+#include <cstring>
+#include <cstdarg>
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+#include "gtest/gtest.h"
+#include "test_system.h"
+
+#include <opae/access.h>
+#include <opae/mmio.h>
+#include <linux/ioctl.h>
+
+using namespace opae::testing;
+
+int mmio_ioctl(mock_object * m, int request, va_list argp) {
+	int retval = -1;
+	errno = EINVAL;
+	UNUSED_PARAM(m);
+	UNUSED_PARAM(request);
+	struct fpga_port_region_info *rinfo = va_arg(argp, struct fpga_port_region_info *);
+	if (!rinfo) {
+		FPGA_MSG("rinfo is NULL");
+		goto out_EINVAL;
+	}
+	if (rinfo->argsz != sizeof(*rinfo)) {
+		FPGA_MSG("wrong structure size");
+		goto out_EINVAL;
+	}
+	if (rinfo->index > 1) {
+		FPGA_MSG("unsupported MMIO index");
+		goto out_EINVAL;
+	}
+	if (rinfo->padding != 0) {
+		FPGA_MSG("unsupported padding");
+		goto out_EINVAL;
+	}
+	rinfo->flags = FPGA_REGION_READ | FPGA_REGION_WRITE | FPGA_REGION_MMAP;
+	rinfo->size = 0x40000;
+	rinfo->offset = 0;
+	retval = 0;
+	errno = 0;
+
+	return retval;
+
+out_EINVAL:
+	retval = -1;
+	errno = EINVAL;
+	return retval;
+}
+
+class opae_feature_enum_c_p : public ::testing::TestWithParam<std::string> {
+	protected:
+	opae_feature_enum_c_p() : filter_(nullptr), tokens_{{nullptr, nullptr}} {}
+
+	virtual void SetUp() override {
+
+		ASSERT_TRUE(test_platform::exists(GetParam()));
+		platform_ = test_platform::get(GetParam());
+		system_ = test_system::instance();
+		system_->initialize();
+		system_->prepare_syfs(platform_);
+		invalid_device_ = test_device::unknown();
+
+		ASSERT_EQ(fpgaInitialize(NULL), FPGA_OK);
+		ASSERT_EQ(fpgaGetProperties(nullptr, &filter_), FPGA_OK);
+		ASSERT_EQ(fpgaPropertiesSetObjectType(filter_, FPGA_ACCELERATOR), FPGA_OK);
+		num_matches_ = 0;
+		ASSERT_EQ(fpgaEnumerate(&filter_, 1, tokens_.data(), tokens_.size(),
+		  &num_matches_), FPGA_OK);
+		EXPECT_EQ(num_matches_, platform_.devices.size());
+		accel_ = nullptr;
+		ASSERT_EQ(fpgaOpen(tokens_[0], &accel_, 0), FPGA_OK);
+		system_->register_ioctl_handler(FPGA_PORT_GET_REGION_INFO, mmio_ioctl);
+		which_mmio_ = 0;
+		uint64_t *mmio_ptr = nullptr;
+		EXPECT_EQ(fpgaMapMMIO(accel_, which_mmio_, &mmio_ptr), FPGA_OK);
+		EXPECT_NE(mmio_ptr, nullptr);
+
+		feature_filter_.type = FPGA_DMA_FEATURE;
+		fpga_guid guid = {0xE7, 0xE3, 0xE9, 0x58, 0xF2, 0xE8, 0x73, 0x9D, 
+					0xE0, 0x4C, 0x48, 0xC1, 0x58, 0x69, 0x81, 0x87 };
+		memcpy_s(feature_filter_.guid, sizeof(fpga_guid), guid, sizeof(fpga_guid));
+	}
+
+	void DestroyTokens() {
+		for (auto &t : tokens_) {
+			if (t) {
+				EXPECT_EQ(fpgaDestroyToken(&t), FPGA_OK);
+				t = nullptr;
+			}
+		}
+		num_matches_ = 0;
+	}
+
+	virtual void TearDown() override {
+		DestroyTokens();
```

Comment:
Need to unmap MMIO and fpgaClose the accel_ here, or else these are leaked.
Also, need to initialize feature_filter_ to nullptr, and then call appropriate Destroy fn in TearDown if non-null, otherwise it is leaked.